### PR TITLE
[SNOW-901393] External plugins loaded from sys.path

### PR DESCRIPTION
**Pull request by sfc-gh-pjob**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added new automated tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
* Added loading of external plugins installable by simple `pip install` command (plugins need to be available at SnowCLI's `sys.path`)
* Added `snowcli.api.Api` with a goal to make it a class gathering all interfaces which will be implemented in `snowcli.app` package. Plugins should use interfaces, not their implementations.
* Added `snowcli.api.plugin.plugin_config` with `PluginConfig` and `PluginConfigProvider` interface allowing plugins to access their configuration.
* Changed plugin entrypoint group namespace from `snowcli.command` to `snowflake.cli.plugin.command`
* Added new steps required to execute before running integration tests locally (installation of external plugins used in tests)

### Other notes
* **Plugins API is still not ready to use. We need to restructure packages before saying it's ready.**
* I'm not adding release notes in this PR because I'm planning to do it in a PR making changes in the packages structure.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -26,6 +26,14 @@ Every integration test should have `integration` mark. By default, integration t
 
 To execute only integration tests run `pytest -m integration`.
 
+### Installation of external plugins used in integration tests
+
+To run integration tests you have to install external plugins which are part of the repository:
+```bash
+pip install -e test_external_plugins/snowpark_hello_single_command
+pip install -e test_external_plugins/multilingual_hello_command_group
+```
```
Comment (sfc-gh-astus) [inline CONTRIBUTING.md:35]: It sounds like these 2 plugins are required for integration tests? Are they?

File: src/snowcli/api/__init__.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class Api:
+    def __init__(
+        self,
+        plugin_config_provider: PluginConfigProvider,
+    ):
+        self.plugin_config_provider = plugin_config_provider
+
+
+class ApiNotInitializedError(RuntimeError):
```
Comment (sfc-gh-astus) [inline src/snowcli/api/__init__.py:14]: Maybe move it to `snowcli/exception.py`?

File: src/snowcli/app/api_impl/plugin/plugin_config_provider_impl.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+from typing import List
+
+from snowcli.api import PluginConfigProvider
+from snowcli.api.plugin.plugin_config import PluginConfig
+from snowcli.config import cli_config
+
+
+class PluginConfigProviderImpl(PluginConfigProvider):
+    def get_enabled_plugin_names(self) -> List[str]:
+        config_path = ["snowcli", "plugins"]
```
Comment (sfc-gh-astus) [inline src/snowcli/app/api_impl/plugin/plugin_config_provider_impl.py:10]: Maybe change it to class constant?

File: tests_integration/test_external_plugins.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+from tests_integration.testing_utils.assertions.test_result_assertions import (
+    assert_that_result_is_successful,
+    assert_that_result_is_successful_and_output_json_contains,
+)
+from tests_integration.snowflake_connector import snowflake_session
+
+
+@pytest.mark.integration
+def test_loading_of_installed_plugins_if_all_plugins_enabled(runner, snowflake_session):
+    runner.use_config("config_with_enabled_all_external_plugins.toml")
```
Comment (sfc-gh-astus) [inline tests_integration/test_external_plugins.py:12]: Maybe instead of overriding default config, create new `invoke` where you can pass path to config?

Comment (sfc-gh-pjob) [reply tests_integration/test_external_plugins.py:12]: I started with that approach but then I noticed that I am limited to only one version of `invoke` and I have to duplicate code.

The way implemented it allows me to use every version of `invoke` and makes the usage less complitated.

File: src/snowcli/api/__init__.py (hunk 2 above)
Comment (sfc-gh-pjob) [reply src/snowcli/api/__init__.py:14]: `api` package shouldn't use classes/functions placed outside of it

File: CONTRIBUTING.md (hunk 1 above)
Comment (sfc-gh-pjob) [reply CONTRIBUTING.md:35]: They are required (as we discussed today).

**Review by sfc-gh-astus — APPROVED** [review]

File: src/snowcli/app/api_impl/plugin/plugin_config_provider_impl.py (hunk 3 above)
Comment (sfc-gh-pjob) [reply src/snowcli/app/api_impl/plugin/plugin_config_provider_impl.py:10]: Discussed and agreed that current version is ok.

**Pull request merged**